sycl/source/exception.cpp

```diff
@@ -10,9 +10,96 @@
+    : exception(EC, nullptr, Msg) {}
+
+exception::exception(std::error_code EC, const std::string &Msg)
+    : exception(EC, nullptr, Msg) {}
+
+// new SYCL 2020 constructors
+exception::exception(std::error_code EC) : exception(EC, nullptr, "") {}
+
+exception::exception(int EV, const std::error_category &ECat,
+                     const std::string &WhatArg)
+    : exception({EV, ECat}, nullptr, WhatArg) {}
+
+exception::exception(int EV, const std::error_category &ECat,
+                     const char *WhatArg)
+    : exception({EV, ECat}, nullptr, std::string(WhatArg)) {}
+
+exception::exception(int EV, const std::error_category &ECat)
+    : exception({EV, ECat}, nullptr, "") {}
+
+exception::exception(context Ctx, std::error_code EC,
+                     const std::string &WhatArg)
+    : exception(EC, std::make_shared<context>(Ctx), WhatArg) {}
+
+exception::exception(context Ctx, std::error_code EC, const char *WhatArg)
+    : exception(Ctx, EC, std::string(WhatArg)) {}
+
+exception::exception(context Ctx, std::error_code EC)
+    : exception(Ctx, EC, "") {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat,
+                     const char *WhatArg)
+    : exception(Ctx, {EV, ECat}, std::string(WhatArg)) {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat,
+                     const std::string &WhatArg)
+    : exception(Ctx, {EV, ECat}, WhatArg) {}
+
+exception::exception(context Ctx, int EV, const std::error_category &ECat)
+    : exception(Ctx, EV, ECat, "") {}
+
+// protected base constructor for all SYCL 2020 constructors
+exception::exception(std::error_code EC, std::shared_ptr<context> SharedPtrCtx,
+                     const std::string &WhatArg)
+    : MMsg(WhatArg + reserved_for_errorcode), MCLErr(PI_INVALID_VALUE),
+      MContext(SharedPtrCtx) {
+  // For compatibility with previous implementation, we are "hiding" the
+  // std::error_code in the MMsg string, behind the null string terminator
+  int stringTermPoint = MMsg.length() - strlen(reserved_for_errorcode);
+  char *reservedPtr = &MMsg[stringTermPoint];
+  reservedPtr[0] = '\0';
+  reservedPtr++;
+  // insert error code
+  std::error_code *ecPtr = reinterpret_cast<std::error_code *>(reservedPtr);
+  *ecPtr = EC;
+}
+
+const std::error_code &exception::code() const noexcept {
+  const char *whatStr = MMsg.c_str();
+  // advance to inner string-terminator
+  int stringTermPoint = MMsg.length() - strlen(reserved_for_errorcode);
+  if (stringTermPoint >= 0) {
+    const char *reservedPtr = &whatStr[stringTermPoint];
+    // check for string terminator, which denotes a SYCL 2020 exception
+    if (reservedPtr[0] == '\0') {
+      reservedPtr++;
+      const std::error_code *ecPtr =
+          reinterpret_cast<const std::error_code *>(reservedPtr);
+      return *ecPtr;
+    }
+  }
+  // else the exception originates from some SYCL 1.2.1 source
+  return sycl121_proxy_errorcode;
```

Comment:
Could you clarify the use case when this function can be called for an exception from SYCL 1.2.1 source, please?
If it's possible, we can not use `MMsg` message string to store the error code. 
Any SYCL 1.2.1 exception with message satisfying following limitations will be recognized as SYCL 2020 exception:
- message string length >= `reserved_for_errorcode` length
- `\0` is located at length of `reserved_for_errorcode` before the end of the message string 

According to my understanding both conditions above is better to replace with assertions.

Reply:
In their simplest form the exceptions newly specified for SYCL 2020 introduce ABI/API breaking changes. The goal here is to support SYCL 2020 exceptions but preserving the structure of the SYCL 1.2.1 ones for compatibility. For example, if a shared library written with SYCL 1.2.1 throws an exception, it should function even if the app that catches it is using SYCL 2020 exceptions.  I don't think we want to assert.

You are right, if a SYCL 1.2.1 exception happens to put a string terminator into the middle of the string at this exact position, then it would be mistakenly recognized as a SYCL 2020 exception. I originally had a 'SYCL' marker inserted to help with forward string scanning, but was asked to remove it. But if restored then we face another hypothetical: "What if a SYCL 1.2.1 exception inserts a null terminator AND the 'SYCL' marker?".   

Ultimately, I doubt anyone is inserting null string terminators into their messages.

Reply:
Ultimately, if we are hiding this data behind the terminator, there are three ways of finding it:

- forward scan for the terminator.  Problem: std::string can contain null terminators.
- reverse scan for the terminator.  Problem: the error_code will most certainly contain bytes of value 0
- locate by position, since the size of the error code is fixed.

The third one is what we are using now, it seems the safest and is simple.  However, if we are worried about it, it might be slightly improved upon by using a multi-byte marker of some sort.  For example: `\0\0\0\0` or `\0SYCL` or a combination of both: `\0S\0Y\0C\0L\0F\0T\0W`

Reply:
I think this limitation is significant enough to be mentioned as a known issue in the documentation.

Reply:
@bader  - do you have a recommendation on where it should be documented, exactly?  I was hoping to just add this note to one of the existing docs, but unsure which is the best fit.  Maybe  sycl/doc/FAQ.md ?

Reply:
https://github.com/intel/llvm/blob/sycl/sycl/doc/GetStartedGuide.md#known-issues-and-limitations is a good place to start.